### PR TITLE
Shrink the code size for debug

### PR DESCRIPTION
**Pull request by davidchisnall**
This is two parts (in two commits)

The first copies the existing debug infrastructure into the loader and strips out the bits that the loader doesn't need.  The second rewrites the generic debug infrastructure to do type erasure in inline functions and then call into a library function for the real work.  This removes the big template instantiations from most callers.

### CODE REVIEW

**Review by nwf-msr — COMMENTED** [review]
Are we sure we don't leak the UART MMIO cap back to the caller, since we're going via libcall?  If we aren't sure, that might not be a blocker, but we should have it documented somewhere that whatever UART is used by the debug library must not be trusted for other purposes.

**Comment by davidchisnall** [comment]
> Are we sure we don't leak the UART MMIO cap back to the caller, since we're going via libcall? If we aren't sure, that might not be a blocker, but we should have it documented somewhere that whatever UART is used by the debug library must not be trusted for other purposes.

Without leaking the UART, a caller is able to disable interrupts and write an arbitrary amount to the UART.  If we leak the UART, there isn't really anything else that they can do.  Moving the UART access into this library isn't a security thing.

File: sdk/lib/debug/debug.cc [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+			write(' ');
+			write(C.base());
+			write('-');
+			write(C.top());
+			write(" l:");
+			write(C.length());
+			write(" o:");
+			write(C.type());
+			write(" p: ");
+			write(C.permissions());
```
[the remark this answers is not in the document]
Comment (rmn30) [reply sdk/lib/debug/debug.cc:99]: Missing close bracket.

**Comment by davidchisnall** [comment]
Creating the outer structure containing multiple pointers for each element ended up introducing a lot of large stack allocations.  I've refactored it so that the interface to the library is less generic, but lets us just pass most of the values that we need in registers.  We still need two pointers of stack space for each format argument.

**Comment by davidchisnall** [comment]
The new version should now flip CI failures, failing in Sail and passing in Ibex.  Once cheriot-sail#14 is merged and we update the dev container, it should pass everywhere.

Size of code + data for the test suite shrinks from 126912 B -> 94304 B.

**Pull request merged**